components/butterfly/src/rumor/dat_file.rs

```diff
@@ -0,0 +1,361 @@
+// Copyright (c) 2017 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fs::{self, File, OpenOptions};
+use std::io::{self, Read, Seek, SeekFrom, Write};
+use std::mem;
+use std::path::{Path, PathBuf};
+
+use byteorder::{ByteOrder, LittleEndian};
+use protobuf::{self, Message};
+
+use error::{Result, Error};
+use member::{Health, Member, MemberList};
+use message::swim::Membership as ProtoMembership;
+use rumor::{Election, ElectionUpdate, Rumor, RumorStore, Service, ServiceConfig, ServiceFile};
+use server::Server;
+
+const HEADER_VERSION: u8 = 1;
+
+/// A versioned binary file containing rumors exchanged by the butterfly server which have
+/// been periodically persisted to disk.
+///
+/// The contents of the DatFile can be used to rebuild and resume a butterfly server connection
+/// if it has been destroyed or lost.
+///
+/// * Header Version - 1 byte
+/// * Header Body - Variable bytes - see Header
+/// * Rumors - Variable bytes
+#[derive(Debug)]
+pub struct DatFile {
+    header: Header,
+    path: PathBuf,
+}
+
+impl DatFile {
+    pub fn new<T: AsRef<Path>>(member_id: &str, data_path: T) -> Self {
+        DatFile {
+            path: data_path.as_ref().join(format!("{}.rst", member_id)),
+            header: Header::default(),
+        }
+    }
+
+    pub fn path(&self) -> &Path {
+        &self.path
+    }
+
+    pub fn read_into(&mut self, server: &Server) -> Result<()> {
+        let mut version = [0; 1];
+        let mut size_buf = [0; 8];
+        // JW: Resizing this buffer is terrible for performance, but it's the easiest way to
+        // read exactly N bytes from a file. I'm not sure what the right approach is but this
+        // won't be a performance issue for a long time anyway, if ever.
+        let mut rumor_buf: Vec<u8> = vec![];
+        let mut bytes_read = 0;
+        let mut file =
+            File::open(&self.path).map_err(|err| Error::DatFileIO(self.path.clone(), err))?;
```

Comment:
Did you consider wrapping this in a `BufReader`? I'm not as sure about the performance of `Read.read_exact` vs. `ButReader.read_exact` but the low level `Read` can make more raw sys calls than its buffered cousin.